### PR TITLE
[JENKINS-46257] Improve Run Parameters folders support for auto-completion

### PR DESCRIPTION
**Pull request by neaguandrei101**
See [JENKINS-46257](https://issues.jenkins-ci.org/browse/JENKINS-46257).

Now the autocompletion works with folders, it suggests the correct relative names of the project from the root or the current context. It behaves exactly like in https://github.com/jenkinsci/parameterized-trigger-plugin/blob/master/src/main/java/hudson/plugins/parameterizedtrigger/BuildTriggerConfig.java#L799.

Previous closed PR https://github.com/jenkinsci/jenkins/pull/3083

If this is merged the same improvements can be made to hudson.tasks.BuildTrigger

I could not write unit tests for this but it is tested manually.

### Proposed changelog entries

* Improve Run Parameters folders support for auto-completion

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

File: core/src/main/java/hudson/model/AutoCompletionCandidates.java
Comment (res0nance) [inline core/src/main/java/hudson/model/AutoCompletionCandidates.java]: Why are we not using the old function?

File: core/src/main/java/hudson/model/RunParameterDefinition.java
Comment (res0nance) [inline core/src/main/java/hudson/model/RunParameterDefinition.java]: No star imports

**Review by res0nance — CHANGES_REQUESTED** [review]
Could you also add a test

File: core/src/main/java/hudson/model/AutoCompletionCandidates.java
Comment (neaguandrei101) [reply core/src/main/java/hudson/model/AutoCompletionCandidates.java]: There are 3 ways autocomplete was implemented
The oldest and worst - RunParameterDefinition - It has no folder support or context
Better - BuildTrigger - It has no folder support but it has context
The newest in Parametrized Build Trigger - It has folder support and context
In the previous PR segalaj, the commiter wanted to implement the 2nd option. Oleg said that it is not enough, it must support folder. I wrote the new function because the old one does not support folder suggestions and this one does. I got inspired from the 3rd option.

File: core/src/main/java/hudson/model/RunParameterDefinition.java
Comment (neaguandrei101) [reply core/src/main/java/hudson/model/RunParameterDefinition.java]: I fixed this in the latest commit.

**Comment by neaguandrei101** [comment]
I have written a unit test based on the old one. In the test you can see the functionality of the new function, folders and context. I have also done extensive manual tests.

**Comment by neaguandrei101** [comment]
If the features in the method do not match what is wanted write exactly what auto completion should be suggested so I can modify the method and test

File: core/src/main/java/hudson/model/AutoCompletionCandidates.java
Comment (rsandell) [inline core/src/main/java/hudson/model/AutoCompletionCandidates.java]: `getAllItems` can be quite expensive to run.

File: test/src/test/java/hudson/model/AutoCompletionCandidatesTest.java
Comment (alecharp) [inline test/src/test/java/hudson/model/AutoCompletionCandidatesTest.java]: This is not testing the new method for the relative path and I'm not confident that the new method is working properly for the relative paths. Could you please use the new method signature to validate the relative path? Only the new one is used in the `doAutoComplete*` method, so we might have a feature regression on the form.

**Review by alecharp — CHANGES_REQUESTED** [review]
I'm not confident that the auto completion will work with relative path.

File: test/src/test/java/hudson/model/AutoCompletionCandidatesTest.java
Comment (neaguandrei101) [reply test/src/test/java/hudson/model/AutoCompletionCandidatesTest.java]: You are right. That part is clearly my mistake. I will fix that part of the test. I got a comment that ```getAllItems()``` is expensive to run but I really need to use it. I am not experienced enough to do something  more complex than that.

**Comment by neaguandrei101** [comment]
In my latest commit I fixed my test mistake as @alecharp pointed out. I want to tell you the following, in the previous test,```completion()``` inside AutoCompletionCandidatesTest these lines pass. 
```
c = AutoCompletionCandidates.ofJobNames(Item.class, "/", x3, x3.getParent());
        assertContains(c, "/foo", "/bar");
```
```
 c = AutoCompletionCandidates.ofJobNames(Item.class, "/bar/", x3, x3.getParent());
        assertContains(c, "/bar/x=1", "/bar/x=2", "/bar/x=3");
```
When I try to manual test the autocompletion results using absolute paths like in the test I get nothing. Also if I manually put the path starting with "/" in the box I get an error. This tells me that absolute paths should not be used because they do not work. What I understand is that "/bar/" is wrong and it should be "bar/" in that case. My new method has relative paths from the root, not starting with "/" and it works. Please tell me if my method is correct, also inspect the test that I wrote for my new method to see if it looks right. In my head everything checks out.

**Comment by neaguandrei101** [comment]
Also I realize my git commits look messy. Could you please point me out on some guidelines to make your jobs easier. I never worked with git on a professional level.

**Comment by daniel-beck** [comment]
@neaguandrei101 No worries, once this is ready it's going to be merged as a single commit. As long as https://github.com/jenkinsci/jenkins/pull/4551/files looks reasonable, it should be fine. Don't rewrite history, as that makes reviewing more difficult.

File: core/src/main/java/hudson/model/AutoCompletionCandidates.java
Comment (res0nance) [reply core/src/main/java/hudson/model/AutoCompletionCandidates.java]: This might be a good candidate for getAllItems(Class, Predicate). It does iterate every job but it checks the predicate condition before permissions. Note getAllItems and friends already does a hasPermission check for Item.READ so the additional check below is not required.

File: core/src/main/java/hudson/model/RunParameterDefinition.java
Comment (res0nance) [inline core/src/main/java/hudson/model/RunParameterDefinition.java]: Does it not work to just do `AutoCompletionCandidates.ofJobNames(Job.class, value, context);` and leverage existing code?

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
I agree with @res0nance that we could significantly improve performance by using the new Predicate API. We could actually improve it even more if the cycle was exiting after reaching the desired number of candidates. I believe that limiting to 10..20 would be more than enough for UX.

**Comment by oleg-nenashev** [comment]
@neaguandrei101 hi, are you still interested in this PR?

**Comment by neaguandrei101** [comment]
I am sorry for being inactive. I have been working hard in a non programming job. I wish to start working on this once school starts. It has been on my mind for a while...

**Comment by alecharp** [comment]
@neaguandrei101 let us know if you feel you can work on this in a near future or if we should close the PR? Thanks.

**Comment by neaguandrei101** [comment]
> @neaguandrei101 let us know if you feel you can work on this in a near future or if we should close the PR? Thanks.

Working on it.

**Comment by neaguandrei101** [comment]
Not good yet. It needs more refinement.

**Comment by alecharp** [comment]
@neaguandrei101 do you know when you will have time to come back on this PR? Thanks.

**Comment by alecharp** [comment]
@neaguandrei101 I've added the `proposed-for-close` label here has we haven't seen activities here since October 2020. Please let us know if you plan to work on this. If the PR is closed in the meantime, we can always reopen it.

**Comment by neaguandrei101** [comment]
Hello,

Please close it. I can't really work on it right now.

Kind Regards

**Comment by alecharp** [comment]
@neaguandrei101 do you know when you would be able to work on this again? It was 18months since your previous comment. Closing this PR wouldn't trash your initial work, and the PR could be reopen when you have time to focus back on it.

**Comment by basil** [comment]
> Please close it. I can't really work on it right now.

Request granted. On behalf of the core team, I would like to thank you for your contribution. Even though this PR did not make it across the finish line, it was a promising start! I continue to encourage you (or anyone else who is interested!) to pick up this PR and drive it to completion. Thanks!

**Comment by neaguandrei101** [comment]
I will try to work on it this summer. I would love to find a solution which
works best. I just was and am quite busy with survival.

**Pull request closed without merging**
